extensions/notebook/src/dialog/managePackages/managePackagesDialogModel.ts

```diff
@@ -0,0 +1,204 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { JupyterServerInstallation } from '../../jupyter/jupyterServerInstallation';
+import { IPackageManageProvider, IPackageDetails } from '../../types';
+
+export interface ManagePackageDialogOptions {
+	multiLocations: boolean;
+	defaultLocation?: string;
+}
+
+/**
+ * Manage package dialog model
+ */
+export class ManagePackageDialogModel {
+
+	public currentProvider: string;
+	private _locations: Set<string> = new Set<string>();
+	private _packageTypes: Map<string, IPackageManageProvider[]> = new Map<string, IPackageManageProvider[]>();
+
+	/**
+	 * Creates new instance of the model
+	 * @param _jupyterInstallation Jupyter installation
+	 * @param _packageManageProviders package manage providers
+	 * @param _options dialog options
+	 */
+	constructor(
+		private _jupyterInstallation: JupyterServerInstallation,
+		private _packageManageProviders: Map<string, IPackageManageProvider>,
+		private _options?: ManagePackageDialogOptions) {
+
+		if (!this._packageManageProviders || this._packageManageProviders.size === 0) {
+			throw Error('Invalid list of package manager providers');
+		}
+	}
+
+	/**
+	 * Initialized the model
+	 */
+	public async init(): Promise<void> {
+		await this.loadCaches();
+		this.loadOptions();
+	}
+
+	/**
+	 * Loads the model options
+	 */
+	private loadOptions(): void {
+		// Set Default Options
+		//
+		if (!this._options) {
+			this._options = {
+				multiLocations: true,
+				defaultLocation: undefined
+			};
+		}
+
+		if (this._options.defaultLocation && !this._packageTypes.get(this._options.defaultLocation)) {
+			throw new Error(`Invalid default location '${this._options.defaultLocation}`);
+		}
+
+		if (!this._options.multiLocations && !this.defaultLocation) {
+			throw new Error('Default location not specified for single location mode');
+		}
+	}
+
+	/**
+	 * Returns the providers map
+	 */
+	public get packageManageProviders(): Map<string, IPackageManageProvider> {
+		return this._packageManageProviders;
+	}
+
+	/**
+	 * Returns the current provider
+	 */
+	public get currentPackageManageProvider(): IPackageManageProvider {
+		if (this.currentProvider) {
+			let provider = this._packageManageProviders.get(this.currentProvider);
+			return provider;
+		}
+		return undefined;
+	}
+
+	/**
+	 * Returns true if multi locations mode is enabled
+	 */
+	public get multiLocationMode(): boolean {
+		return this._options.multiLocations;
+	}
```

Comment:
_options might be undefined. Would be good to have places in here that use that handle that case appropriately